exchanges/coinbene/coinbene_websocket.go

```diff
@@ -0,0 +1,305 @@
+package coinbene
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/gorilla/websocket"
+	"github.com/thrasher-corp/gocryptotrader/common"
+	"github.com/thrasher-corp/gocryptotrader/currency"
+	exchange "github.com/thrasher-corp/gocryptotrader/exchanges"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/orderbook"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/websocket/wshandler"
+	"github.com/thrasher-corp/gocryptotrader/exchanges/websocket/wsorderbook"
+	log "github.com/thrasher-corp/gocryptotrader/logger"
+)
+
+const (
+	coinbeneWsURL = "wss://ws-contract.coinbene.vip/openapi/ws"
```

Comment:
```
[ERROR]: 2019/10/17 13:07:31 routines.go exchange Coinbene websocket error - Message: unsupported topic orderBook.ABTETH.100. Code: 10503
[ERROR]: 2019/10/17 13:07:31 routines.go exchange Coinbene websocket error - Message: unsupported topic orderBook.ABTUSDT.100. Code: 10503
[ERROR]: 2019/10/17 13:07:31 routines.go exchange Coinbene websocket error - Message: unsupported topic orderBook.ADIETH.100. Code: 10503
[ERROR]: 2019/10/17 13:07:31 routines.go exchange Coinbene websocket error - Message: unsupported topic orderBook.ADKBTC.100. Code: 10503
[ERROR]: 2019/10/17 13:07:31 routines.go exchange Coinbene websocket error - Message: unsupported topic orderBook.ADNBTC.100. Code: 10503
[ERROR]: 2019/10/17 13:07:31 routines.go exchange Coinbene websocket error - Message: unsupported topic orderBook.AIDBTC.100. Code: 10503
[ERROR]: 2019/10/17 13:07:31 routines.go exchange Coinbene websocket error - Message: unsupported topic orderBook.AITUSDT.100. Code: 10503
[ERROR]: 2019/10/17 13:07:31 routines.go exchange Coinbene websocket error - Message: unsupported topic orderBook.ALIETH.100. Code: 10503
[ERROR]: 2019/10/17 13:07:31 routines.go exchange Coinbene websocket error - Message: unsupported topic orderBook.ALXETH.100. Code: 10503
[ERROR]: 2019/10/17 13:07:31 routines.go exchange Coinbene websocket error - Message: unsupported topic orderBook.APLETH.100. Code: 10503
[ERROR]: 2019/10/17 13:07:31 routines.go exchange Coinbene websocket error - Message: unsupported topic orderBook.ATXBTC.100. Code: 10503
[ERROR]: 2019/10/17 13:07:31 routines.go exchange Coinbene websocket error - Message: unsupported topic orderBook.B2GBTC.100. Code: 10503
[ERROR]: 2019/10/17 13:07:31 routines.go exchange Coinbene websocket error - Message: unsupported topic orderBook.B91USDT.100. Code: 10503
[ERROR]: 2019/10/17 13:07:31 routines.go exchange Coinbene websocket error - Message: unsupported topic orderBook.BATBTC.100. Code: 10503
[ERROR]: 2019/10/17 13:07:31 routines.go exchange Coinbene websocket error - Message: unsupported topic orderBook.BEZBTC.100. Code: 10503
[ERROR]: 2019/10/17 13:07:31 routines.go exchange Coinbene websocket error - Message: unsupported topic orderBook.BGCUSDT.100. Code: 10503
[ERROR]: 2019/10/17 13:07:31 routines.go exchange Coinbene websocket error - Message: unsupported topic orderBook.BKGBTC.100. Code: 10503
[ERROR]: 2019/10/17 13:07:31 routines.go exchange Coinbene websocket error - Message: unsupported topic orderBook.BNTBTC.100. Code: 10503
[ERROR]: 2019/10/17 13:07:31 routines.go exchange Coinbene websocket error - Message: unsupported topic orderBook.BOAUSDT.100. Code: 10503
[ERROR]: 2019/10/17 13:07:31 routines.go exchange Coinbene websocket error - Message: unsupported topic orderBook.BTTUSDT.100. Code: 10503
[ERROR]: 2019/10/17 13:07:31 routines.go exchange Coinbene websocket error - Message: unsupported topic orderBook.BVTETH.100. Code: 10503
[ERROR]: 2019/10/17 13:07:31 routines.go exchange Coinbene websocket error - Message: unsupported topic orderBook.C3WETH.100. Code: 10503
[ERROR]: 2019/10/17 13:07:31 routines.go exchange Coinbene websocket error - Message: unsupported topic orderBook.CANETH.100. Code: 10503
[ERROR]: 2019/10/17 13:07:31 routines.go exchange Coinbene websocket error - Message: unsupported topic orderBook.CCCETH.100. Code: 10503
[ERROR]: 2019/10/17 13:07:31 routines.go exchange Coinbene websocket error - Message: unsupported topic orderBook.CCEUSDT.100. Code: 10503
[ERROR]: 2019/10/17 13:07:31 routines.go exchange Coinbene websocket error - Message: unsupported topic orderBook.CFTUSDT.100. Code: 10503
[ERROR]: 2019/10/17 13:07:31 routines.go exchange Coinbene websocket error - Message: unsupported topic orderBook.CLOBTC.100. Code: 10503
[ERROR]: 2019/10/17 13:07:31 routines.go exchange Coinbene websocket error - Message: unsupported topic orderBook.CMTETH.100. Code: 10503
[ERROR]: 2019/10/17 13:07:31 routines.go exchange Coinbene websocket error - Message: unsupported topic orderBook.CMTUSDT.100. Code: 10503
[ERROR]: 2019/10/17 13:07:31 routines.go exchange Coinbene websocket error - Message: unsupported topic orderBook.CNNBTC.100. Code: 10503
[ERROR]: 2019/10/17 13:07:31 routines.go exchange Coinbene websocket error - Message: unsupported topic orderBook.CNNETH.100. Code: 10503
[ERROR]: 2019/10/17 13:07:31 routines.go exchange Coinbene websocket error - Message: unsupported topic orderBook.CNNUSDT.100. Code: 10503
[ERROR]: 2019/10/17 13:07:31 routines.go exchange Coinbene websocket error - Message: unsupported topic orderBook.CPCBTC.100. Code: 10503
[ERROR]: 2019/10/17 13:07:31 routines.go exchange Coinbene websocket error - Message: unsupported topic orderBook.CRNBTC.100. Code: 10503
[ERROR]: 2019/10/17 13:07:31 routines.go exchange Coinbene websocket error - Message: unsupported topic orderBook.CVCBTC.100. Code: 10503
[ERROR]: 2019/10/17 13:07:31 routines.go exchange Coinbene websocket error - Message: unsupported topic orderBook.CXCUSDT.100. Code: 10503
[ERROR]: 2019/10/17 13:07:31 routines.go exchange Coinbene websocket error - Message: unsupported topic orderBook.CXPBTC.100. Code: 10503
[ERROR]: 2019/10/17 13:07:31 routines.go exchange Coinbene websocket error - Message: unsupported topic orderBook.DCAETH.100. Code: 10503
[ERROR]: 2019/10/17 13:07:31 routines.go exchange Coinbene websocket error - Message: unsupported topic orderBook.DCTBTC.100. Code: 10503
[ERROR]: 2019/10/17 13:07:31 routines.go exchange Coinbene websocket error - Message: unsupported topic orderBook.DGDBTC.100. Code: 10503
[ERROR]: 2019/10/17 13:07:31 routines.go exchange Coinbene websocket error - Message: unsupported topic orderBook.DTAETH.100. Code: 10503
[ERROR]: 2019/10/17 13:07:31 routines.go exchange Coinbene websocket error - Message: unsupported topic orderBook.DUCBTC.100. Code: 10503
[ERROR]: 2019/10/17 13:07:31 routines.go exchange Coinbene websocket error - Message: unsupported topic orderBook.DVCETH.100. Code: 10503
[ERROR]: 2019/10/17 13:07:31 routines.go exchange Coinbene websocket error - Message: unsupported topic orderBook.EBCBTC.100. Code: 10503
[ERROR]: 2019/10/17 13:07:31 routines.go exchange Coinbene websocket error - Message: unsupported topic orderBook.EBCETH.100. Code: 10503
[ERROR]: 2019/10/17 13:07:31 routines.go exchange Coinbene websocket error - Message: unsupported topic orderBook.EBCUSDT.100. Code: 10503
[ERROR]: 2019/10/17 13:07:31 routines.go exchange Coinbene websocket error - Message: unsupported topic orderBook.ECABTC.100. Code: 10503
```

As mentioned before this websocket is not designed for their regular exchange but for 

https://www.coinbene.com/contract/usdt/BTC-SWAP
https://www.coinbene.com/contract/btc/BTCUSDT

How does this impact with usual orderbook/ticker requirements etc? @thrasher- might be able to shed some light on

Reply:
I feel the best approach would be to read the error, determine the subscription and remove it from the subscribed pairs. 

Specific currencies for specific asset types is a feature of `engine` and could be adjust there post engine-merge, or post this branch rebased into engine. Pairs supported can be retrievable via http://openapi-contract.coinbene.com/api/swap/v2/market/tickers in the `FetchTradablePairs` funcs

Reply:
There's currently no way in master to abstract a spot pair from a swap deriv contract. Will make the necessary changes on engine as it supports multiple asset types when merging this in and can leave websocket disabled by default